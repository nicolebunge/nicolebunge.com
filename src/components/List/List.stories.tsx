import List from '.';

const meta = {
  component: List,
};

interface Item {
  id: number;
  name: string;
}

const items = [
  { id: 1, name: 'One' },
  { id: 2, name: 'Two' },
  { id: 3, name: 'Three' },
] as const satisfies Item[];

const Default = {
  args: {
    items,
    renderItem: (item: Item) => item.name,
  },
};

export { Default };
export default meta;
