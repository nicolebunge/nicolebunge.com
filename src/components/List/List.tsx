import type { Key, ReactNode } from 'react';

interface ListProps<Item> {
  as?: 'ul' | 'ol';
  items: Item[];
  renderItem: (item: Item) => ReactNode;
}

function List<Item extends { id: Key }>(props: ListProps<Item>) {
  const { as: As = 'ul', items, renderItem } = props;

  return (
    <As>
      {items.map((item) => {
        const { id } = item;

        return <li key={id}>{renderItem(item)}</li>;
      })}
    </As>
  );
}

export type { ListProps };
export default List;
