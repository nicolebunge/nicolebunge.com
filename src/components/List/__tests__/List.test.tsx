import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import * as stories from '../List.stories';

const { Default } = composeStories(stories);

describe('List component', () => {
  it('renders correctly', () => {
    const { container } = render(<Default />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
