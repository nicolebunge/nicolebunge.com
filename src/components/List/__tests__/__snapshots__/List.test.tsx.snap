// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List component > renders correctly 1`] = `
<ul>
  <li>
    One
  </li>
  <li>
    Two
  </li>
  <li>
    Three
  </li>
</ul>
`;
